src/auth_action.rs

```diff
@@ -44,78 +46,57 @@ impl AuthAction {
         //todo(adam-cattermole):hostvar resolver?
         // store dynamic metadata in filter state
         debug!("process_response(auth): store_metadata");
-        if store_metadata(check_response.get_dynamic_metadata()).is_err() {
-            return self.resolve_failure_mode();
-        }
+        store_metadata(check_response.get_dynamic_metadata())
+            .map_err(ProcessGrpcMessageError::Property)?;
 
         match check_response.http_response {
             None => {
                 debug!("process_response(auth): received no http_response");
-                self.resolve_failure_mode()
+                Err(ProcessGrpcMessageError::EmptyResponse)
             }
             Some(CheckResponse_oneof_http_response::denied_response(denied_response)) => {
                 debug!("process_response(auth): received DeniedHttpResponse");
-                let status_code = denied_response.get_status().get_code();
-                let response_headers = Self::get_header_vec(denied_response.get_headers());
-                Err(GrpcErrResponse::new(
-                    status_code as u32,
-                    response_headers,
-                    denied_response.body,
-                ))
+                let direct_response: DirectResponse = denied_response.into();
+                Ok(direct_response.into())
             }
             Some(CheckResponse_oneof_http_response::ok_response(ok_response)) => {
                 debug!("process_response(auth): received OkHttpResponse");
 
                 if !ok_response.get_response_headers_to_add().is_empty() {
                     warn!("process_response(auth): Unsupported field 'response_headers_to_add' in OkHttpResponse");
-                    return Err(GrpcErrResponse::new_internal_server_error());
+                    Err(ProcessGrpcMessageError::UnsupportedField)
                 } else if !ok_response.get_headers_to_remove().is_empty() {
                     warn!("process_response(auth): Unsupported field 'headers_to_remove' in OkHttpResponse");
-                    return Err(GrpcErrResponse::new_internal_server_error());
+                    Err(ProcessGrpcMessageError::UnsupportedField)
                 } else if !ok_response.get_query_parameters_to_set().is_empty() {
                     warn!("process_response(auth): Unsupported field 'query_parameters_to_set' in OkHttpResponse");
-                    return Err(GrpcErrResponse::new_internal_server_error());
+                    Err(ProcessGrpcMessageError::UnsupportedField)
                 } else if !ok_response.get_query_parameters_to_remove().is_empty() {
                     warn!("process_response(auth): Unsupported field 'query_parameters_to_remove' in OkHttpResponse");
-                    return Err(GrpcErrResponse::new_internal_server_error());
+                    Err(ProcessGrpcMessageError::UnsupportedField)
+                } else {
```

Comment:
Does this still enforce an error or does this take into account `FailureMode` now? these cases should all error regardless of `FailureMode`

Reply:
grpc response processing result is checked in `src/runtime_action.rs` for all type of actions. Currently, `FailureMode` is applied for `ProcessGrpcMessageError::UnsupportedField` and any other error. 

iiuc, currently, when processing the grpc response, errors from parsing the response bytes (ProtobufError) can be bypassed by FailureMode but the actual errors processing by rate limit action or auth action should all fail regardless of FailureMode. Is that correct? Can I ask what is the rationale for doing that? why differentiation?

I am happy to keep that behavior, it's two liner change. But I want to understand the differences :pray: .

Reply:
I'm unsure exactly for the ratelimit action, I believe the behaviour was replicating the existing behaviour. For these cases in the auth action; the fields are unsupported by the external auth service and could be an indication of a man in the middle attack, so the request should not proceed

Reply:
@adam-cattermole I added a [commit](https://github.com/Kuadrant/wasm-shim/pull/196/commits/bb77ba497d3caed9f77a5e8fec3ebad99bd8bc6c) addressing the comment. Hopefully it is like you asked.